File: src/clewin/Camera.java
package clewin;

/**
 * (c) Clewin Griffith 2013
 */
public class Camera {

	private Point location;
	private double fov;
	private Point target;
	private double focalLength; //distance to screen
	private Direction up = new Direction(0.0, 1.0, 0.0);
	
	public Camera locatedAt(Point p) {
		this.location = p;
		return this;
	}
	
	public Camera withFieldOfView(double fovDegrees)
	{
		this.fov = Math.toRadians(fovDegrees);
		return this;
	}
	
	public Camera withFocalLength(double len) {
		this.focalLength = len;
		return this;
	}
	
	public Camera pointingAt(Point p) {
		this.target = p;
		return this;
	}
	
	public Camera withUpVector(Direction up) {
		this.up = up;
		return this;
	}
	
	public double getFov() {
		return this.fov;
	}
	
	public Point getLocation() {
		return this.location;
	}

	public double getFocalLength() {
		return this.focalLength;
	}
	
	public Direction getLookVector() {
		return Vector.direction(location, target);
	}
	
	public Direction getUpVector() {
		return up;
	}
	
}

File: src/clewin/Camera.scala
package clewin

/**
 * (c) Clewin Griffith 2013
 */
class Camera {
  def locatedAt(p: Point): Camera = {
    this.location = p
    return this
  }

  def withFieldOfView(fovDegrees: Double): Camera = {
    this.fov = Math.toRadians(fovDegrees)
    return this
  }

  def withFocalLength(len: Double): Camera = {
    this.focalLength = len
    return this
  }

  def pointingAt(p: Point): Camera = {
    this.target = p
    return this
  }

  def withUpVector(up: Vec3): Camera = {
    this.up = up
    return this
  }

  def getFov: Double = {
    return this.fov
  }

  def getLocation: Point = {
    return this.location
  }

  def getFocalLength: Double = {
    return this.focalLength
  }

  def getLookVector: Vec3 = {
    return Util.direction(location, target)
  }

  def getUpVector: Vec3 = {
    return up
  }

  private var location: Point = null
  private var fov: Double = .0
  private var target: Point = null
  private var focalLength: Double = .0
  private var up: Vec3 = new Vec3(0.0, 1.0, 0.0)
}